maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,320 @@
+        Situation 3: Return all PMs with enough resources but not empty.
+
+        Args: vm_req_cpu (int): The CPU resource of the VM requirement.
+        """
+        # NOTE: Should we implement this logic inside the action scope?
+        # TODO: In oversubscribable scenario, we should consider more situations, like
+        #       the PM type (oversubscribable and non-oversubscribable).
+
+        valid_pm_list = []
+        for pm in self._machines:
+            if pm.req_cpu == 0:
+                valid_pm_list.append(
+                    ValidPhysicalMachine(pm_id=pm.id, remaining_cpu=pm.cap_cpu, remaining_mem=pm.cap_mem)
+                )
+                break
+            elif pm.req_cpu > 0 and (pm.cap_cpu - pm.req_cpu) >= vm_req_cpu:
+                valid_pm_list.append(ValidPhysicalMachine(
+                    pm_id=pm.id,
+                    remaining_cpu=pm.cap_cpu - pm.req_cpu,
+                    remaining_mem=pm.cap_mem - pm.req_mem
+                ))
+        return valid_pm_list
+
+    def _on_vm_required(self, vm_requirement_event: CascadeEvent):
+        """Callback when there is a VM requirement generated."""
+        # Get VM data from payload.
+        payload: VmRequirementPayload = vm_requirement_event.payload
+
+        vm_req: VirtualMachine = payload.vm_req
+        remaining_buffer_time: int = payload.remaining_buffer_time
+
+        self._pending_vm_req_payload[vm_req.id] = payload
+
+        valid_pm_list = self._get_valid_pm(vm_req.req_cpu)
+
+        if len(valid_pm_list) > 0:
+            # Generate pending decision.
+            decision_payload = DecisionPayload(
+                valid_pm=valid_pm_list,
+                vm_id=vm_req.id,
+                vm_req_cpu=vm_req.req_cpu,
+                vm_req_mem=vm_req.req_mem,
+                remaining_buffer_time=remaining_buffer_time
+            )
+            pending_decision_event = self._event_buffer.gen_decision_event(
+                tick=vm_requirement_event.tick, payload=decision_payload)
+            vm_requirement_event.add_immediate_event(event=pending_decision_event)
+        else:
+            # Either postpone the requirement event or failed.
+            self._postpone_vm_requirement(
+                postpone_type=PostponeType.Resource,
+                vm_id=vm_req.id,
+                remaining_buffer_time=remaining_buffer_time
+            )
+
+    def _on_vm_finished(self, evt: AtomEvent):
+        """Callback when there is a VM ready to be terminated."""
+
+        # Get the VM info.
+        payload: VmFinishedPayload = evt.payload
+        vm_id = payload.vm_id
+        vm: VirtualMachine = self._live_vm[vm_id]
+
+        # Release PM resources.
+        pm: PhysicalMachine = self._machines[vm.pm_id]
+        pm.req_cpu -= vm.req_cpu
+        pm.req_mem -= vm.req_mem
+        # Calculate the PM's utilization.
+        pm_util_cpu = (pm.cap_cpu * pm.util_cpu - vm.req_cpu * vm.util_cpu) / pm.cap_cpu
```

Comment:
refine the pm.update_workload interface, which may support finished_vms, updated_vms, new_coming_vms?